vsintegration/src/FSharp.Editor/FSharpBraceMatchingService.fs

```diff
@@ -0,0 +1,115 @@
+
+open Microsoft.VisualStudio.FSharp.LanguageService
+open Microsoft.VisualStudio.Text
+open Microsoft.VisualStudio.Text.Tagging
+
+open Microsoft.FSharp.Compiler.Parser
+open Microsoft.FSharp.Compiler.SourceCodeServices
+
+[<ExportBraceMatcher(FSharpCommonConstants.FSharpLanguageName)>]
+type internal FSharpBraceMatchingService() =
+              
+    static member private SupportedBraceTypes = [
+        ('(', ')');
+        ('<', '>');
+        ('[', ']');
+        ('{', '}');
+    ]
+    
+    static member private IgnoredClassificationTypes = [
+        ClassificationTypeNames.Comment;
+        ClassificationTypeNames.StringLiteral;
+        ClassificationTypeNames.ExcludedCode;
+    ]
+
+    interface IBraceMatcher with
+        member this.FindBracesAsync(document: Document, position: int, cancellationToken: CancellationToken): Task<Nullable<BraceMatchingResult>> =
+            let computation = async {
+                let! sourceText = document.GetTextAsync(cancellationToken) |> Async.AwaitTask
+                return
+                    try
+                        FSharpBraceMatchingService.GetBraceMatchingResult(sourceText, document.Name, position, cancellationToken)
+                    with ex -> 
+                        Assert.Exception(ex)
+                        reraise()
+            }
+            Async.StartAsTask(computation, TaskCreationOptions.None, cancellationToken)
+  
+    // Helper function to proxy Roslyn types to tests
+    static member FindMatchingBrace(sourceText: SourceText, fileName: string, position: int, cancellationToken: CancellationToken) : Option<int> =
+        let braceMatchingResult = FSharpBraceMatchingService.GetBraceMatchingResult(sourceText, fileName, position, cancellationToken)
+        if braceMatchingResult.HasValue then
+            if braceMatchingResult.Value.LeftSpan.Start = position then
+                Some(braceMatchingResult.Value.RightSpan.Start)
+            else if braceMatchingResult.Value.RightSpan.Start = position then
+                Some(braceMatchingResult.Value.LeftSpan.Start)
+            else
+                None
+        else
+            None
+
+    static member private GetBraceMatchingResult(sourceText: SourceText, fileName: string, position: int, cancellationToken: CancellationToken) : Nullable<BraceMatchingResult> =
+        if position < 0 || position >= sourceText.Length then
+            Nullable()
+        else
+            let classificationData = FSharpColorizationService.GetColorizationData(sourceText, fileName, [], cancellationToken)
+
+            let shouldBeIgnored(characterPosition) =
+                match classificationData.GetClassifiedSpan(characterPosition) with
+                | None -> false
+                | Some(classifiedSpan) -> FSharpBraceMatchingService.IgnoredClassificationTypes |> Seq.contains classifiedSpan.ClassificationType
+                
+            if shouldBeIgnored(position) then
+                Nullable()
```

Comment:
In F# logic, use "Some/None", and convert to a nullable at the interop boundary.